src/main/scala/io/iohk/ethereum/network/rlpx/RLPxConnectionHandler.scala

```diff
@@ -136,21 +138,53 @@ class RLPxConnectionHandler(
         context.parent ! MessageReceived(message)
 
       case Failure(ex) =>
-        log.error(ex, "Cannot decode message")
+        log.error(ex, s"Cannot decode message from $peerId")
     }
 
-    def handshaked(messageCodec: MessageCodec): Receive =
+    /**
+      * Handles sending and receiving messages from the Akka TCP connection, while also handling acknowledgement of
+      * messages sent. Messages are only sent when all Ack from previous messages were received.
+      *
+      * @param messageCodec
+      * @param messagesNotSent, messages not yet sent
+      * @param messageAwaitingAck, message sent for which we are awaiting an acknowledgement by Akka TCP
+      */
+    def handshaked(messageCodec: MessageCodec,
+                   messagesNotSent: Queue[MessageSerializable] = Queue.empty,
+                   messageAwaitingAck: Option[MessageSerializable] = None): Receive =
       handleWriteFailed orElse handleConnectionClosed orElse {
         case sm: SendMessage =>
-          val out = messageCodec.encodeMessage(sm.serializable)
-          connection ! Write(out)
-          log.debug("Sent message: {}", sm.serializable)
+          if(messageAwaitingAck.isEmpty) {
+            sendMessage(messageCodec, sm.serializable)
+            context become handshaked(messageCodec, messagesNotSent, Some(sm.serializable))
+          } else
+            context become handshaked(messageCodec, messagesNotSent :+ sm.serializable, messageAwaitingAck)
```

Comment:
After further testing, I realized that sending new messages upon `Ack` timeouts so didn't work and caused `Write to ... peer failed` errors. The reason for this was the same as for the problem that is being solved in this PR, which is that a msgs was written to the TCP connection while another one's queueing was in progress (the one we are awaiting for the `Ack`).
We should then close the TCP connection when encountering that an expected `Ack` was not received.